### PR TITLE
HAL_F4Light: latest changes

### PR DESCRIPTION
**Pull request by night-ghost**
also fixes https://github.com/ArduPilot/ardupilot/issues/8300 and https://github.com/ArduPilot/ardupilot/issues/8238

### CODE REVIEW

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp [hunk 1]
```diff
@@ -251,7 +255,7 @@ void AP_Compass_MAG3110::read()
         return;
     }
 
-    Vector3f field(_mag_x_accum, _mag_y_accum, _mag_z_accum);
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_MAG3110.cpp]: did you find this change made a difference? It should be equivalent (and the original code should be very slightly faster)
If this did change behavior then we need to fix that elsewhere

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp [hunk 2]
```diff
@@ -195,15 +199,15 @@ bool AP_Compass_MAG3110::_read_sample()
 }
 
 
-#define MAG_SCALE 1.0/10000  // 1 Tesla full scale of +-10000
+#define MAG_SCALE (1.0/10000)  // 1 Tesla full scale of +-10000
```
Comment (khancyr) [inline libraries/AP_Compass/AP_Compass_MAG3110.cpp]: could you change to 1.0f to ensure that we are using float calculation

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp [hunk 3]
```diff
@@ -251,7 +255,7 @@ void AP_Compass_MAG3110::read()
         return;
     }
 
-    Vector3f field(_mag_x_accum, _mag_y_accum, _mag_z_accum);
+    Vector3f field = Vector3f(_mag_x_accum, _mag_y_accum, _mag_z_accum);
```
Comment (khancyr) [inline libraries/AP_Compass/AP_Compass_MAG3110.cpp]: this shouldn't matter, but why did you change to the copy constructor ?

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp (hunk 2 above)
Comment (night-ghost) [reply libraries/AP_Compass/AP_Compass_MAG3110.cpp]: dot in 1.0 guarantees that this is float, so 1.0f is excessive tautology, oiled oil.

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp (hunk 3 above)
Comment (night-ghost) [reply libraries/AP_Compass/AP_Compass_MAG3110.cpp]: I got a garbage in 'field' variable before somehow (in debugger). This compiles to pure calculations without any memory copy

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp (hunk 1 above)
Comment (night-ghost) [reply libraries/AP_Compass/AP_Compass_MAG3110.cpp]: yes I think it too but first  code leaves garbage in 'field' variable somehow, checked by debugger. 2nd variant compiles into pure calculations without any memory copy/move.

**Comment by night-ghost** [comment]
seems that I missed something but while driver calls "publish_filtered_field()" regularily, I don't see compass data via MAVlink stream

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp (hunk 1 above)
Comment (night-ghost) [reply libraries/AP_Compass/AP_Compass_MAG3110.cpp]: this may be not GCC issue but seems that Eclipse bug, but I want to see variables

File: libraries/AP_Compass/AP_Compass_MAG3110.cpp (hunk 2 above)
Comment (khancyr) [reply libraries/AP_Compass/AP_Compass_MAG3110.cpp]: 1.0 means double in c++

Comment (night-ghost) [reply libraries/AP_Compass/AP_Compass_MAG3110.cpp]: yes, if there is no key -fsingle-precision-constant in command line. But as you wish

**Comment by night-ghost** [comment]
Driver works now - there was a bad scale
